python: tweak script linking in virtualenv

UniqMartin opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [x] Have you successfully run `brew tests` with your changes locally?

----

**Remaining steps/tasks (see below for context):**

- [ ] Agree on the preferred interface for the non-block form.
- [x] Replace minimal implementation with proper one without block/recursion.
- [x] Update documentation in code (YARD).
- [ ] Update documentation in `share/doc/homebrew/`.
- [x] Update tests.

-----

This is an attempt to refine #344, to address https://github.com/Homebrew/brew/pull/344#issuecomment-232577429 by @MikeMcQuaid:

> > `venv.link_scripts(bin) { venv.pip_install "." }`
>
> Don't understand why `venv.link_scripts` is a block rather than a following command.

@ilovezfs expressed similar concerns in https://github.com/Homebrew/brew/pull/344#issuecomment-234242676:

> Also, it does still seem awkward to have `venv.link_scripts` taking the block that does ` venv.pip_install buildpath`. Could the virtualenv accumulate before and after state without needing to resort to nesting?

----

Instead of making the formula author use a slightly awkward block construct and expose this implementation detail (diffing of `@venv_root/"bin"`), offer a more familiar:

```ruby
venv.pip_install buildpath, :link_scripts => bin
```

The default will be not to link any scripts, but this can be of course reversed if that's a useful default behavior (in this case, the opt-out would be `:link_scripts => false`).

*Note:* The current state of the PR is primarily to solicit feedback whether this is a reasonable change to make to the user-facing API of the `virtualenv` support, assuming my proposal doesn't come a bit late (as far as I'm aware this part of the interface hasn't been used yet in formulae).

cc @tdsmith

MikeMcQuaid commented:
Makes sense to me; can you supply a sample formula diff?

ilovezfs commented:
any thoughts on having install & install_and_link, and exposing even less of this?

tdsmith commented:
This was the first way I implemented it. I came to feel that using an options hash was not as nice as composing methods, based on a prejudice against arguments that switch behavior on or off; you can find some comments in that PR where I vacillate between the two options.

> install & install_and_link

This was the second way I implemented it! I think this is probably nicer than an options hash.

UniqMartin commented:
Pushed a few updates to make the code changes more complete.

> Makes sense to me; can you supply a sample formula diff?

Most formulae will use the single-shot wrapper `virtualenv_install_with_resources` that already exists. The more complicated cases will use `virtualenv_create` and operate on the returned `Virtualenv` instance. For these, the only difference will be to replace:

```
venv.link_scripts(bin) { venv.pip_install buildpath }
```

With:

```ruby
venv.pip_install buildpath, :link_scripts => bin
```

> any thoughts on having install & install_and_link, and exposing even less of this?

I'm fine with that and it would be a trivial change. Basically, I think either of the following read nicer than the block form (though I'm personally most fond of the option hash form):

```ruby
venv.pip_install_and_link buildpath
venv.pip_install buildpath, :link_scripts => true
venv.pip_install buildpath, :link_scripts => bin
```

`Virtualenv#pip_install_and_link` would be basically the current `Virtualenv#link_scripts` but instead of `yield`ing, it would forward its arguments to `Virtualenv#pip_install`.

UniqMartin commented:
Replaced the functionality of `:link_scripts => bin` with a `#pip_install_and_link` method that uses composition instead and delegates to `#pip_install` for installing the targets.

tdsmith commented on Library/Homebrew/test/test_language_python.rb:
```diff
@@ -71,18 +73,24 @@ def test_pip_install_accepts_resource
-    refute((bin/"kilroy").exist?)
-    refute((dest/"kilroy").exist?)
+  def test_pip_install_and_link_links_scripts
+    bin = (@dir/"bin").tap(&:mkpath)
```
What does this do?

UniqMartin replied:
Single-line version of:

```ruby
```

tdsmith replied:
Can we spell that out? I think the reference to `tap` is confusing.

UniqMartin replied:
I was trying to avoid repeating `bin` again and used something every Ruby object has. I used (or maybe abused) `Object#tap`.

UniqMartin replied:
> Can we spell that out?

Yes, we can!

tdsmith replied:
I learned a thing today!

tdsmith commented on Library/Homebrew/test/test_language_python.rb:
```diff
@@ -71,18 +73,24 @@ def test_pip_install_accepts_resource
-    assert((dest/"kilroy").exist?)
-    assert((dest/"kilroy").symlink?)
+    @formula.expects(:system).returns(true).with do |*params|
+      FileUtils.touch bin/"kilroy"
```
The `with` block instructs the mock object about how we expect it should be called; the block should return true if the call is what we expected and shouldn't have any side effects.

UniqMartin replied:
Can you suggest the correct way how I can make the underlying `pip_install` invocation produce a side effect? Aside from that, I guess in this test I should be doing something like `@venv.expects(:pip_install).yields(nil)` instead …

tdsmith commented on Library/Homebrew/test/test_language_python.rb:
```diff
@@ -71,18 +73,24 @@ def test_pip_install_accepts_resource
+    assert_predicate dest/"kilroy", :symlink?
     assert_equal((bin/"kilroy").realpath, (dest/"kilroy").realpath)
-    refute((dest/"irrelevant").exist?)
+    refute_predicate dest/"irrelevant", :exist?
```
Not related, but why do we prefer this style?

UniqMartin replied:
Because it results in a clearer message if the assertion fails. Instead of just saying “assertion failed” (no message is provided) it dumps the `Pathname` instance and the predicate of that instance that was supposed to be true/false.

UniqMartin commented on Library/Homebrew/test/test_language_python.rb:
```diff
@@ -71,18 +73,24 @@ def test_pip_install_accepts_resource
-    refute((dest/"kilroy").exist?)
+  def test_pip_install_and_link_links_scripts
+    bin = (@dir/"bin").tap(&:mkpath)
+    dest = @formula.dest
```
Indeed. That's a stupid last-minute change I made because I was shuffling variable names back and forth and sadly I failed to run the test suite afterwards.

UniqMartin commented:
I have addressed all comments. I'm not sure how reasonable the stuff is, that I'm doing in the test, but at least I'm not sticking things with side effects into places where they don't belong.

tdsmith commented on Library/Homebrew/test/test_language_python.rb:
```diff
@@ -71,18 +73,27 @@ def test_pip_install_accepts_resource
+    bin_before = Dir[bin/"*"]
+    FileUtils.touch bin/"kilroy"
+    bin_after = Dir[bin/"*"]
+    @venv.expects(:pip_install).with { |*params| params == ["foo"] }
```
You can just say `.with("foo")`: http://gofreerange.com/mocha/docs/Mocha/Expectation.html#with-instance_method

UniqMartin replied:
Thanks! I didn't realize this could be simplified like that for simple cases like this one.

tdsmith commented:
:+1:

UniqMartin commented:
Apologies that I was hardly around the past few days!

I have to admit that I'm not entirely happy that this was merged without completing the checklist in the top comment (e.g. update documentation in `share/doc/homebrew/`) and with all the noise of the incremental changes preserved in the commit message …
